### PR TITLE
Move cargo tests to compiletest.

### LINKED ISSUE #177: Port Cargo Tests into compiletest.

### PR DESCRIPTION
**Pull request by bdalrhm**
### Description of changes: 

This PR moves cargo tests to compiletest. `expected.*` files are renamed to `*.expected` to make use of available convenience functions. Note that a `--target-dir` option is added to change the default `target` directories for cargo builds to `build/<arch>/test/cargo/<proj-name>/<function-name>/target` and avoid race problems while running cargo tests concurrently. For consistency, other generated files are also moved into the specified target directory.

### Resolved issues:

Resolves #177

### Testing:

* How is this change tested?
  * Cargo tests still pass after the changes
* Is this a refactor change?
  * To some degree, `cargo-rmc-tests` is moved and renamed to `src/test/cargo`

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/tools/compiletest/src/main.rs [hunk 1]
```diff
@@ -582,6 +582,24 @@ fn collect_tests_from_dir(
     let build_dir = output_relative_path(config, relative_dir_path);
     fs::create_dir_all(&build_dir).unwrap();
 
+    // If we find a `Cargo.toml` file in the current directory and we're in
+    // Cargo mode, we should look for `*.expected` files and create an output
+    // directory corresponding to each to avoid race conditions during the
+    // testing phase. We immedietly return after adding the tests to avoid
```
Comment (adpaco-aws) [inline src/tools/compiletest/src/main.rs:588]:
```suggestion
    // testing phase. We immediately return after adding the tests to avoid
```

File: cargo-rmc-tests/.gitignore [hunk 2]
```diff
@@ -1,7 +0,0 @@
-empty-main/Cargo.lock
```
Comment (adpaco-aws) [inline cargo-rmc-tests/.gitignore:1]: This file appears as removed and then added somewhere else, instead of just moved. Because of that, the copyright check is failing.

File: cargo-rmc-tests/empty-main/expected.main [hunk 3]
```diff
@@ -1 +0,0 @@
-VERIFICATION SUCCESSFUL
```
Comment (adpaco-aws) [inline cargo-rmc-tests/empty-main/expected.main:1]: Same here, should have been moved. This is not a big deal, but if you could change so that we avoid overriding the copyright check, that would be great.

File: scripts/rmc.py [hunk 4]
```diff
@@ -102,11 +102,11 @@ def run_cbmc(cbmc_filename, cbmc_args, verbose=False, quiet=False):
...
-def run_visualize(cbmc_filename, cbmc_args, verbose=False, quiet=False, keep_temps=False, function="main", srcdir=".", wkdir="."):
-    results_filename = "results.xml"
-    coverage_filename = "coverage.xml"
-    property_filename = "property.xml"
-    temp_goto_filename = "temp.goto"
+def run_visualize(cbmc_filename, cbmc_args, verbose=False, quiet=False, keep_temps=False, function="main", srcdir=".", wkdir=".", outdir="."):
+    results_filename = os.path.join(outdir, "results.xml")
+    coverage_filename = os.path.join(outdir, "coverage.xml")
+    property_filename = os.path.join(outdir, "property.xml")
+    temp_goto_filename = os.path.join(outdir, "temp.goto")
```
Comment (adpaco-aws) [inline scripts/rmc.py:109]: We should involve @vecchiot-aws in this PR to check at least edits to this file.

File: src/tools/compiletest/src/runtest.rs [hunk 5]
```diff
@@ -2411,6 +2414,47 @@ impl<'test> TestCx<'test> {
         }
     }
 
+    /// Runs cargo-rmc on the function specified by the stem of `self.testpaths.file`.
+    /// An error message is printed to stdout if verification result does not
+    /// contain the expected output in `self.testpaths.file`.
```
Comment (adpaco-aws) [inline src/tools/compiletest/src/runtest.rs:2419]: Great comments, thanks so much!

File: src/test/cargo/simple-main/main.expected [hunk 6]
```diff
@@ -0,0 +1 @@
+[main.assertion.1] line 4 assertion failed: 1 == 2: FAILURE
```
Comment (adpaco-aws) [inline src/test/cargo/simple-main/main.expected:1]: Why did you change the expected output from this test?

File: src/tools/compiletest/src/common.rs [hunk 7]
```diff
@@ -24,6 +24,7 @@ pub enum Mode {
     MirOpt,
     Assembly,
     RMC,
+    Cargo,
```
Comment (adpaco-aws) [inline src/tools/compiletest/src/common.rs:27]: Can this be `CargoRMC` instead?

File: src/tools/compiletest/src/common.rs [hunk 8]
```diff
@@ -55,6 +56,7 @@ impl FromStr for Mode {
             "mir-opt" => Ok(MirOpt),
             "assembly" => Ok(Assembly),
             "rmc" => Ok(RMC),
+            "cargo" => Ok(Cargo),
```
Comment (adpaco-aws) [inline src/tools/compiletest/src/common.rs:59]: Same here, can we call them `cargo-rmc`?

**Review by adpaco-aws — CHANGES_REQUESTED** [review]
This looks great! Could we use `CargoRMC` and `cargo-rmc` for the enum variant and its associated string?

File: scripts/cargo-rmc [hunk 9]
```diff
@@ -25,6 +26,8 @@ def main():
     parser.add_argument("--mangler", default="v0")
     parser.add_argument("--visualize", action="store_true")
     parser.add_argument("--srcdir", default=".")
+    parser.add_argument("--target-dir", default="target",
```
Comment (vecchiot-aws) [inline scripts/cargo-rmc:29]: I don't know how to comment on a non-changed file, but should we be adding this argument to `scripts/rmc` as well?

Comment (bdalrhm) [reply scripts/cargo-rmc:29]: I thought about that, but I could not come to a conclusion. Normally, when someone runs `cargo`, they expect all artifacts (besides `Cargo.lock`) to be in the target directory. `rustc` has a somewhat similar option `--out-dir` for specifying the directory of the compiled file. However, it's not really treated as a build directory and I don't know if we should treat it as such. So, I think this is a decision for the team to make.

File: src/test/cargo/simple-main/main.expected (hunk 6 above)
Comment (bdalrhm) [reply src/test/cargo/simple-main/main.expected]: I did so because I noticed that the current cargo-rmc tests do not check for the behavior of `cargo-rmc` when a test fails. I modified this test with a simple assertion to ensure that
1. `cargo-rmc` does not accept all inputs.
2. Fails for "right" reasons.

File: cargo-rmc-tests/empty-main/expected.main (hunk 3 above)
Comment (bdalrhm) [reply cargo-rmc-tests/empty-main/expected.main:1]: See the [comment below](https://github.com/model-checking/rmc/pull/198#discussion_r651023636) for why I modified this file. If I only move this `empty-main` test, then I will have to create another test (and another `*.expected` file) to check for failure and face the same copyright problem again. So I think it's better to make a decision on this problem with this PR.

File: cargo-rmc-tests/.gitignore (hunk 2 above)
Comment (bdalrhm) [reply cargo-rmc-tests/.gitignore:1]: Hmm... I can do that. But then we will have to go through another PR to modify the contents.

File: scripts/cargo-rmc (hunk 9 above)
Comment (adpaco-aws) [reply scripts/cargo-rmc:29]: In addition to @bdalrhm explanation, I would like to point out that RMC's workflow is more similar to CBMC's: It does generate a verification output, not a compiled program. So it makes sense for us to have this `--target-dir` flag for `cargo-rmc` but not for `rmc` itself.

**Review by adpaco-aws — APPROVED** [review]
